### PR TITLE
rgw: fix forward request for bulkupload to be applied in multisite

### PR DESCRIPTION
**Pull request by Jing-Scott**
fixes: http://tracker.ceph.com/issues/19645
Signed-off-by: Jing Wenjun <jingwenjun@cmss.chinamobile.com>

### CODE REVIEW

**Comment by Jing-Scott** [comment]
@rzarzynski @cbodley can you help to take a look? Thanks:)

File: src/rgw/rgw_op.cc [hunk 1]
```diff
@@ -5580,6 +5580,18 @@ int RGWBulkUploadOp::handle_dir_verify_permission()
   return 0;
 }
 
+void RGWBulkUploadOp::forward_req_info(req_info& info, std::string bucket_name)
+{
+  if (info.script_uri.find(bucket_name) != std::string::npos) {
```
Comment (cbodley) [inline src/rgw/rgw_op.cc]: in what cases would this already contain the bucket name? maybe add a comment here to clarify?

File: src/rgw/rgw_op.cc [hunk 2]
```diff
@@ -5580,6 +5580,18 @@ int RGWBulkUploadOp::handle_dir_verify_permission()
   return 0;
 }
 
+void RGWBulkUploadOp::forward_req_info(req_info& info, std::string bucket_name)
```
Comment (cbodley) [inline src/rgw/rgw_op.cc]: this doesn't need to be a member of `RGWBulkUploadOp` - can you change it to a static function (and add a `CephContext *cct` argument for the `ldout`)?

`bucket_name` should be passed as a const reference

**Comment by Jing-Scott** [comment]
@cbodley done!

**Comment by cbodley** [comment]
@rzarzynski can you help me figure out how to test this one?

**Comment by rzarzynski** [comment]
> can you help me figure out how to test this one?

@cbodley: we would need two things:
1. operational multi-site deployment,
2. uncompressed TAR archive for bulk upload.

@Jing-Scott provided `curl` examples in PR https://github.com/ceph/ceph/pull/14775.

**Comment by cbodley** [comment]
tested a bulk upload to a secondary zone, and verified that the create_bucket requests were correctly forwarded and processed on the master zone 👍

**Comment by rzarzynski** [comment]
This branch has been tested in following Teuthology runs:
* http://pulpito.ceph.com/rzarzynski-2017-04-29_14:39:05-rgw-wip-rzarzynski-testing---basic-mira/,
* http://pulpito.ceph.com/rzarzynski-2017-04-30_12:33:39-rgw-wip-rzarzynski-testing---basic-mira/.

The results look good. The failures have been caused by the well-known `strange leak of std::string memory from md_config_t seen in radosgw` signalised by Valgrind.

**Pull request merged**